### PR TITLE
Remove instance keys and explicit splatting everywhere

### LINKED ISSUE #1014: Batch selections should not preview single-element non-splats
Although it's not easy to do with our APIs at the moment, it's theoretically possible to log a non-splatted element with a single instance-key.

For example, a future API that allows us to log colors directly:
```
rr.log_points("points", [[1,2],[3,4]], instance_keys=[1, 2])
rr.log_colors("points", [RED], instance_keys=[2])
```

In this case, color is not a splat, but if you hover on the *batch* of points in the UI, it will display as if RED applies to the whole batch rather than only applying to the 2nd point.

### LINKED ISSUE #1777: datastore: implement shared slices for autogenerated keys
suggested by @emilk in https://github.com/rerun-io/rerun/pull/1727#discussion_r1156813046

### LINKED ISSUE #1793: datastore: don't dump autogenerated keys
`DataStore::to_data_tables` should detect and filter out instance keys that were auto-generated when dumping.

### LINKED ISSUE #1893: Eliminate legacy splats
Splats used to require a dedicated row with a special `InstanceKey` and associated machinery.

With the new store, anything that doesn't match the explicit number of instances given by the row should be treated as a splat:
- within a row with `num_instances == 0`, a component of unit length is a splat
- within a row with `num_instances == 1`, a component of unit length is _not_ a splat
- within a row with `num_instances == N` (`N > 1`), a component of unit length is a splat

Note: the renderer uses the old `InstanceKey::SPLAT` constant to drive some related (but different) logic in the selection/hovering subsystems.

Fixing this will fix #1014 in the process.

### LINKED ISSUE #5303: Remove instance keys (Archetype-less, PoV-less, join-less queries)
Random thoughts off the top of my head:
- Remove `InstanceKey` from logging APIs
- Rework `re_data_store` accordingly:
  - No auto-generated instance keys
  - No clustering key checks
  - Probably lots of tests to fix
- Rework `re_query` accordingly:
  - Make it archetype-less, key-less, PoV-less, join-less
  - Do we still need `re_query` at all? If we do, it can be _vastly_ simplified
  - Probably lots of tests to fix
- Rework `re_query_cache` accordingly:
  - Make it archetype-less, key-less, PoV-less, join-less
  - Probably lots of tests to fix
- Update every downstream crates accordingly:
  - Stuff driven by `InstanceKey` (e.g. hover) should be driven by 
  - Visualizers 
- Introduce a new Rerun-agnostic `Identifier` component
  - Just a u64 for now, until we have datatype conversions
  - Doesn't do any magic, just a way to have persistent IDs to drive visualizer features

This is not an exhaustive list, `InstanceKey`s are everywhere.

It will be way simpler to build all of this on the side while keeping the existing stuff untouched and then switch everything, rather than living across two worlds for who knows how long.
I.e. build all of that stuff bite-size pieces in parallel, and only start removing the old stuff when things are in place.

### LINKED ISSUE #5685: Kiss-ICP with a mem limit is permanently GC-ing because _something_ is hogging memory
Via @02alexander:
image

which basically means that we're GC-ing like crazy in hopes of reducing memory, except the thing that's actually hogging memory is not part of the GC path.

Hence the question is: what is the something in question? :eye:

### PR DESCRIPTION
**Pull request by teh-cmc**
Annihilate everything related to instance keys.

---

- [x] Remove `InstanceKey`
- [x] Remove `NumInstances`
- [x] Remove explicit/client-side splats
- [x] Remove cluster keys and autogenerated cells
- [x] Clean up every other little thing related to any of the above
- [x] Update batching docs
- [x] Check that Kiss-ICP now works smoothly
- [x] Check roundtrips
- [x] Check `rerun snippets/*.rrd`
- [x] Rebase and run checklist
- [x] Run full check bot

---

Findings (none blocking, not sure any are new -- to be investigated later):
- Depth clouds with visible history are broken in a different way than before
- `OutOfTreeTransform` might be broken? (see associated snippet)
- Does `DisconnectedSpace` even work? (see associated snippet)
- Is something wrong with labels when there are lots of instances?

---

- Fixes #5303
- Fixes #1014
- Fixes #1777
- Fixes #1793
- Fixes #1893
- Fixes #5685
- Fixes #1014

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6104?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6104?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!

- [PR Build Summary](https://build.rerun.io/pr/6104)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

**Comment by teh-cmc** [comment]
@rerun-bot full-check

**Comment by teh-cmc** [comment]
@rerun-bot full-check

**Comment by teh-cmc** [comment]
@rerun-bot full-check

File: crates/re_log_types/src/lib.rs [hunk 1]
```diff
@@ -83,6 +82,90 @@ macro_rules! impl_into_enum {
 
 // ----------------------------------------------------------------------------
 
+/// A unique numeric index for each individual instance within a batch.
+///
+/// Use [`Instance::ALL`] to refer to all instances in a batch.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct Instance(pub(crate) u64);
```
Comment (emilk) [inline crates/re_log_types/src/lib.rs:90]: nit: this deserves its own file imho

File: crates/re_log_types/src/lib.rs [hunk 2]
```diff
@@ -83,6 +82,90 @@ macro_rules! impl_into_enum {
...
+}
+
+impl std::fmt::Display for Instance {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.is_all() {
+            "<all>".fmt(f)
+        } else {
+            // TODO(emilk): re_format::uint(self.0).fmt(f) (fix cyclic dependency!)
+            self.0.fmt(f)
```
Comment (emilk) [inline crates/re_log_types/src/lib.rs:162]: it's fixed! we can depend on re_format now!

**Review by emilk — APPROVED** [review]
Amazing - hell of a job

**Pull request merged**